lib/index.ts

```diff
@@ -0,0 +1,244 @@
+    let results: MetricsResults[] = [];
+
+    // Kill spawned Chrome process in case of ctrl-C.
+    process.on(_SIGINT, async () => {
+      if (this.launcher) {
+        await this.launcher.kill();
+        process.exit(_SIGINT_EXIT_CODE);
+        console.log(messages.getMessage('CLOSING_CHROME'));
+      }
+    });
+
+    for (let runIndex of runs) {
+      try {
+        results[runIndex] = await this.run();
+        console.log(messages.getMessageWithPrefix('SUCCESS', 'SUCCESS_RUN', runIndex, runs.length));
+      } catch(error) {
+        results[runIndex] = error;
+        console.error(messages.getMessageWithPrefix('ERROR', 'FAILED_RUN', runIndex, runs.length, error.message));
+      }
+    }
+
+    const ret = {runs: results.filter(r => !(r instanceof Error))};
+    if (this.runs > 1 && !this.flags.submit) {
+      const median = this.findMedianRun(results);
+      console.log(messages.getMessage('MEDIAN_RUN'));
+      this.displayOutput(median);
+    } else if (this.flags.submit) {
+      const sheets = new Sheets(this.sheets, this.clientSecret);
+      return sheets.appendResults(results).then(() => results);
+    }
+    return ret;
+  }
+
+  async run(): Promise<MetricsResults> {
+    try {
+      let lhResults: LighthouseResults;
+      await this.launchChrome();
+
+      if (process.env.CI) {
+        // handling CRI_TIMEOUT issue - https://github.com/GoogleChrome/lighthouse/issues/833
+        this.tryLighthouseCounter = 0;
+        lhResults = await this.runLighthouseOnCI();
+      } else {
+        lhResults = await lighthouse(this.url, this.flags);
+      }
+
+      const metricsResults: MetricsResults = await this.recordLighthouseTrace(lhResults);
+      await this.launcher.kill();
+
+      return metricsResults;
+    } catch (error) {
+      throw new Error(error);
+    }
+  }
+
+  async runLighthouseOnCI(): Promise<LighthouseResults> {
+    try {
+      return await lighthouse(this.url, this.flags, perfConfig);
+    } catch(error) {
+      if (error.code === 'CRI_TIMEOUT' && this.tryLighthouseCounter <= _MAX_LIGHTHOUSE_TRIES) {
+        return await this.retryLighthouseOnCI();
+      }
+
+      if (this.tryLighthouseCounter > _MAX_LIGHTHOUSE_TRIES) {
+        throw new Error(messages.getMessage('CRI_TIMEOUT_REJECT'));
+      }
+    }
+  }
+
+  async retryLighthouseOnCI(): Promise<LighthouseResults> {
+    this.tryLighthouseCounter++;
+    console.log(messages.getMessage('CRI_TIMEOUT_RELAUNCH'));
+
+    try {
+      return await this.runLighthouseOnCI();
+    } catch(error) {
+      console.error(error.message);
+      console.error(messages.getMessage('CLOSING_CHROME'));
+      return this.launcher.kill();
+    }
+  }
+
+  async launchChrome() {
+    try {
+      this.launcher = new (ChromeLauncher.ChromeLauncher || ChromeLauncher)();
+      await this.launcher.isDebuggerReady();
+    } catch(e) {
+      console.log(messages.getMessage('LAUNCHING_CHROME'));
```

Comment:
or are you saying...

if debugger is not ready, launch a new chrome so that we can have a debugger?